src/inmanta/data/__init__.py

```diff
@@ -4641,34 +4641,39 @@ async def get_current_resource_state(cls, env: uuid.UUID, rid: ResourceIdStr) ->
         Return the state of the given resource in the latest version of the configuration model
         or None if the resource is not present in the latest version.
         """
-        query = f"""
+        query = """
+            WITH latest_released_version AS (
+                SELECT max(version) AS version
+                FROM configurationmodel
+                WHERE environment=$1 AND released
+            )
             SELECT (
-                SELECT r.status
-                FROM resource r
-                WHERE r.environment=$1
-                    AND r.model=(
-                        SELECT max(c.version) AS version
-                        FROM {ConfigurationModel.table_name()} c
-                        WHERE c.environment=$1 AND c.released
-                    )
-                    AND r.resource_id=$2
-            ) AS status,
-            (
-                SELECT rps.last_non_deploying_status
-                FROM resource_persistent_state rps
-                WHERE rps.environment=$1 AND rps.resource_id=$2
-            ) AS last_non_deploying_status
+                CASE
+                    -- The resource_persistent_state.last_non_deploying_status column is only populated for
+                    -- actual deployment operations to prevent locking issues. This case-statement calculates
+                    -- the correct state from the combination of the resource table and the
+                    -- resource_persistent_state table.
+                    WHEN r.status::text IN('deploying', 'undefined', 'skipped_for_undefined')
+                        -- The deploying, undefined and skipped_for_undefined states are not tracked in the
+                        -- resource_persistent_state table.
+                        THEN r.status::text
+                    WHEN rps.last_deployed_attribute_hash != r.attribute_hash
+                        -- The hash changed since the last deploy -> new desired state
+                        THEN 'available'
```

Comment:
Could this be `unavailable` as well? i.e. do we consider that "deployed" for the sake of the hash update? We may want to simply return `r.status::text` in this case as well, or am I missing something?

Reply:
I think we can return `r.status::text` in this case.